File: tests/wasm/test_kaufmanstop.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('KAUFMANSTOP partial params', () => {

    const { high, low } = testData;


    const result = wasm.kaufmanstop_js(
        new Float64Array(high),
        new Float64Array(low),
        22, 2.0, 'long', 'sma'
    );
    assert.ok(result instanceof Float64Array, 'Result should be Float64Array');
    assert.strictEqual(result.length, high.length);
});

test('KAUFMANSTOP accuracy', async () => {

    const { high, low } = testData;
    const expected = EXPECTED_OUTPUTS.kaufmanstop;

    const result = wasm.kaufmanstop_js(
        new Float64Array(high),
        new Float64Array(low),
        expected.defaultParams.period,
        expected.defaultParams.mult,
        expected.defaultParams.direction,
        expected.defaultParams.maType
    );

    assert.ok(result instanceof Float64Array, 'Result should be Float64Array');
    assert.strictEqual(result.length, high.length);


    const last5 = Array.from(result).slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-1,
        "KAUFMANSTOP last 5 values mismatch"
    );



});

test('KAUFMANSTOP default candles', () => {

    const { high, low } = testData;
    const expected = EXPECTED_OUTPUTS.kaufmanstop;

    const result = wasm.kaufmanstop_js(
        new Float64Array(high),
        new Float64Array(low),
        expected.defaultParams.period,
        expected.defaultParams.mult,
        expected.defaultParams.direction,
        expected.defaultParams.maType
    );
    assert.ok(result instanceof Float64Array, 'Result should be Float64Array');
    assert.strictEqual(result.length, high.length);
});

test('KAUFMANSTOP zero period', () => {

    const high = new Float64Array([10.0, 20.0, 30.0]);
    const low = new Float64Array([5.0, 15.0, 25.0]);

    assert.throws(() => {
        wasm.kaufmanstop_js(high, low, 0, 2.0, 'long', 'sma');
    }, /Invalid period/);
});

test('KAUFMANSTOP period exceeds length', () => {

    const high = new Float64Array([10.0, 20.0, 30.0]);
    const low = new Float64Array([5.0, 15.0, 25.0]);

    assert.throws(() => {
        wasm.kaufmanstop_js(high, low, 10, 2.0, 'long', 'sma');
    }, /Invalid period/);
});

test('KAUFMANSTOP very small dataset', () => {

    const high = new Float64Array([42.0]);
    const low = new Float64Array([41.0]);

    assert.throws(() => {
        wasm.kaufmanstop_js(high, low, 22, 2.0, 'long', 'sma');
    }, /Invalid period|Not enough valid data/);
});

test('KAUFMANSTOP empty data', () => {

    const empty = new Float64Array([]);

    assert.throws(() => {
        wasm.kaufmanstop_js(empty, empty, 22, 2.0, 'long', 'sma');
    }, /Empty data/);
});

test('KAUFMANSTOP mismatched lengths', () => {

    const high = new Float64Array([10.0, 20.0, 30.0]);
    const low = new Float64Array([5.0, 15.0]);

    assert.throws(() => {
        wasm.kaufmanstop_js(high, low, 2, 2.0, 'long', 'sma');
    });
});

test('KAUFMANSTOP all NaN input', () => {

    const allNaNHigh = new Float64Array(100);
    const allNaNLow = new Float64Array(100);
    allNaNHigh.fill(NaN);
    allNaNLow.fill(NaN);

    assert.throws(() => {
        wasm.kaufmanstop_js(allNaNHigh, allNaNLow, 22, 2.0, 'long', 'sma');
    }, /All values are NaN/);
});

test('KAUFMANSTOP NaN handling', () => {

    const { high, low } = testData;
    const expected = EXPECTED_OUTPUTS.kaufmanstop;

    const result = wasm.kaufmanstop_js(
        new Float64Array(high),
        new Float64Array(low),
        expected.defaultParams.period,
        expected.defaultParams.mult,
        expected.defaultParams.direction,
        expected.defaultParams.maType
    );
    assert.strictEqual(result.length, high.length);


    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }


    const warmup = expected.warmupPeriod;
    for (let i = 0; i < warmup && i < result.length; i++) {
        assert(isNaN(result[i]), `Expected NaN at warmup index ${i}`);
    }
});

test('KAUFMANSTOP with short direction', () => {
    const { high, low } = testData;

    const resultLong = wasm.kaufmanstop_js(
        new Float64Array(high.slice(0, 100)),
        new Float64Array(low.slice(0, 100)),
        22, 2.0, 'long', 'sma'
    );
    const resultShort = wasm.kaufmanstop_js(
        new Float64Array(high.slice(0, 100)),
        new Float64Array(low.slice(0, 100)),
        22, 2.0, 'short', 'sma'
    );


    let foundDifference = false;
    const warmup = 43;
    for (let i = warmup; i < 100; i++) {
        if (!isNaN(resultLong[i]) && !isNaN(resultShort[i]) && resultLong[i] !== resultShort[i]) {
            foundDifference = true;
            break;
        }
    }
    assert.ok(foundDifference, "Long and short directions should produce different results");
});

test('KAUFMANSTOP batch single parameter set', () => {

    const { high, low } = testData;
    const expected = EXPECTED_OUTPUTS.kaufmanstop;

    const batchResult = wasm.kaufmanstop_batch_js(
        new Float64Array(high),
        new Float64Array(low),
        22, 22, 0,
        2.0, 2.0, 0.0,
        'long', 'sma'
    );

    assert.ok(batchResult);
    assert.ok(batchResult.values);
    assert.ok(batchResult.combos);
    assert.strictEqual(batchResult.rows, 1);
    assert.strictEqual(batchResult.cols, high.length);


    const singleRow = batchResult.values.slice(0, high.length);
    const last5 = singleRow.slice(-5);
    assertArrayClose(
        last5,
        expected.batchDefaultRow,
        1e-1,
        "KAUFMANSTOP batch single params mismatch"
    );
});

test('KAUFMANSTOP batch multiple periods', () => {

    const { high, low } = testData;
    const testHigh = new Float64Array(high.slice(0, 100));
    const testLow = new Float64Array(low.slice(0, 100));

    const batchResult = wasm.kaufmanstop_batch_js(
        testHigh, testLow,
        20, 24, 2,
        2.0, 2.0, 0.0,
        'long', 'sma'
    );


    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.cols, 100);
    assert.strictEqual(batchResult.values.length, 3 * 100);
    assert.strictEqual(batchResult.combos.length, 3);


    const periods = [20, 22, 24];
    for (let i = 0; i < periods.length; i++) {
        const rowStart = i * 100;
        const rowData = batchResult.values.slice(rowStart, rowStart + 100);

        const singleResult = wasm.kaufmanstop_js(
            testHigh, testLow,
            periods[i], 2.0, 'long', 'sma'
        );

        assertArrayClose(
            rowData,
            singleResult,
            1e-10,
            `Period ${periods[i]} mismatch`
        );
    }
});

test('KAUFMANSTOP batch metadata', () => {

    const { high, low } = testData;
    const testHigh = new Float64Array(high.slice(0, 50));
    const testLow = new Float64Array(low.slice(0, 50));

    const result = wasm.kaufmanstop_batch_js(
        testHigh, testLow,
        20, 22, 2,
        1.5, 2.0, 0.5,
        'long', 'sma'
    );


    assert.strictEqual(result.combos.length, 4);
    assert.strictEqual(result.rows, 4);
    assert.strictEqual(result.cols, 50);


    assert.strictEqual(result.combos[0].period, 20);
    assert.strictEqual(result.combos[0].mult, 1.5);
    assert.strictEqual(result.combos[0].direction, 'long');
    assert.strictEqual(result.combos[0].ma_type, 'sma');


    assert.strictEqual(result.combos[3].period, 22);
    assert.strictEqual(result.combos[3].mult, 2.0);
});

test('KAUFMANSTOP batch edge cases', () => {

    const testData = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25]);


    const singleBatch = wasm.kaufmanstop_batch_js(
        testData, testData,
        5, 5, 1,
        2.0, 2.0, 0.1,
        'long', 'sma'
    );

    assert.strictEqual(singleBatch.values.length, 25);
    assert.strictEqual(singleBatch.combos.length, 1);


    const largeBatch = wasm.kaufmanstop_batch_js(
        testData, testData,
        5, 7, 10,
        2.0, 2.0, 0,
        'long', 'sma'
    );


    assert.strictEqual(largeBatch.values.length, 25);
    assert.strictEqual(largeBatch.combos.length, 1);
});

test('KAUFMANSTOP different MA types', () => {

    const { high, low } = testData;
    const testHigh = new Float64Array(high.slice(0, 100));
    const testLow = new Float64Array(low.slice(0, 100));
    const maTypes = ['sma', 'ema', 'wma', 'smma'];
    const results = [];

    for (const maType of maTypes) {
        try {
            const result = wasm.kaufmanstop_js(
                testHigh, testLow,
                22, 2.0, 'long', maType
            );
            results.push({ maType, result });
        } catch (e) {

        }
    }


    assert(results.length >= 1, "At least SMA should be supported");


    if (results.length > 1) {
        for (let i = 1; i < results.length; i++) {
            let foundDifference = false;
            for (let j = 21; j < 100; j++) {
                if (!isNaN(results[0].result[j]) && !isNaN(results[i].result[j]) &&
                    Math.abs(results[0].result[j] - results[i].result[j]) > 1e-10) {
                    foundDifference = true;
                    break;
                }
            }


            if (!foundDifference) {
                console.log(`Note: ${results[0].maType} and ${results[i].maType} produced same results`);
            }
        }
    }
});


test('KAUFMANSTOP zero-copy API', () => {
    const high = new Float64Array([100, 102, 101, 103, 102, 104, 103, 105, 104, 106,
                                    105, 107, 106, 108, 107, 109, 108, 110, 109, 111,
                                    110, 112, 111, 113, 112]);
    const low = new Float64Array([99, 101, 100, 102, 101, 103, 102, 104, 103, 105,
                                   104, 106, 105, 107, 106, 108, 107, 109, 108, 110,
                                   109, 111, 110, 112, 111]);
    const len = high.length;


    const highPtr = wasm.kaufmanstop_alloc(len);
    const lowPtr = wasm.kaufmanstop_alloc(len);
    const outPtr = wasm.kaufmanstop_alloc(len);

    assert(highPtr !== 0, 'Failed to allocate high buffer');
    assert(lowPtr !== 0, 'Failed to allocate low buffer');
    assert(outPtr !== 0, 'Failed to allocate output buffer');

    try {

        const highView = new Float64Array(wasm.__wasm.memory.buffer, highPtr, len);
        const lowView = new Float64Array(wasm.__wasm.memory.buffer, lowPtr, len);
        const outView = new Float64Array(wasm.__wasm.memory.buffer, outPtr, len);


        highView.set(high);
        lowView.set(low);


        wasm.kaufmanstop_into(highPtr, lowPtr, outPtr, len, 5, 2.0, 'long', 'sma');


        const regularResult = wasm.kaufmanstop_js(high, low, 5, 2.0, 'long', 'sma');


        const outView2 = new Float64Array(wasm.__wasm.memory.buffer, outPtr, len);

        for (let i = 0; i < len; i++) {
            if (isNaN(regularResult[i]) && isNaN(outView2[i])) {
                continue;
            }
            assert(Math.abs(regularResult[i] - outView2[i]) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${outView2[i]}`);
        }
    } finally {

        wasm.kaufmanstop_free(highPtr, len);
        wasm.kaufmanstop_free(lowPtr, len);
        wasm.kaufmanstop_free(outPtr, len);
    }
});

test('KAUFMANSTOP zero-copy error handling', () => {

    assert.throws(() => {
        wasm.kaufmanstop_into(0, 0, 0, 10, 22, 2.0, 'long', 'sma');
    }, /Null pointer/);


    const ptr = wasm.kaufmanstop_alloc(10);
    try {

        assert.throws(() => {
            wasm.kaufmanstop_into(ptr, ptr, ptr, 10, 0, 2.0, 'long', 'sma');
        }, /Invalid period/);
    } finally {
        wasm.kaufmanstop_free(ptr, 10);
    }
});

test('KAUFMANSTOP memory management', () => {

    const sizes = [100, 1000, 10000];

    for (const size of sizes) {
        const highPtr = wasm.kaufmanstop_alloc(size);
        const lowPtr = wasm.kaufmanstop_alloc(size);
        assert(highPtr !== 0, `Failed to allocate high buffer of ${size} elements`);
        assert(lowPtr !== 0, `Failed to allocate low buffer of ${size} elements`);


        const highView = new Float64Array(wasm.__wasm.memory.buffer, highPtr, size);
        const lowView = new Float64Array(wasm.__wasm.memory.buffer, lowPtr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            highView[i] = i * 1.5;
            lowView[i] = i * 1.2;
        }


        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(highView[i], i * 1.5, `High memory corruption at index ${i}`);
            assert.strictEqual(lowView[i], i * 1.2, `Low memory corruption at index ${i}`);
        }


        wasm.kaufmanstop_free(highPtr, size);
        wasm.kaufmanstop_free(lowPtr, size);
    }
});

test.after(() => {
    console.log('KAUFMANSTOP WASM tests completed');
});

File: tests/python/test_kaufmanstop.py
"""
Python binding tests for KAUFMANSTOP indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestKaufmanstop:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_kaufmanstop_accuracy(self, test_data):
        """Test KAUFMANSTOP matches expected values from Rust tests - mirrors check_kaufmanstop_accuracy"""
        expected = EXPECTED_OUTPUTS['kaufmanstop']
        
        result = ta_indicators.kaufmanstop(
            test_data['high'], 
            test_data['low'], 
            period=expected['default_params']['period'],
            mult=expected['default_params']['mult'],
            direction=expected['default_params']['direction'],
            ma_type=expected['default_params']['ma_type']
        )
        
        assert len(result) == len(test_data['high'])
        
        
        assert_close(
            result[-5:],
            expected['last_5_values'],
            rtol=1e-6,
            msg="KAUFMANSTOP last 5 values mismatch"
        )
        
        
        
    
    def test_kaufmanstop_zero_period(self):
        """Test that kaufmanstop fails with zero period"""
        high = np.array([10.0, 20.0, 30.0])
        low = np.array([5.0, 15.0, 25.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.kaufmanstop(high, low, period=0)
    
    def test_kaufmanstop_period_exceeds_length(self):
        """Test that kaufmanstop fails when period exceeds data length"""
        high = np.array([10.0, 20.0, 30.0])
        low = np.array([5.0, 15.0, 25.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.kaufmanstop(high, low, period=10)
    
    def test_kaufmanstop_mismatched_lengths(self):
        """Test that kaufmanstop fails when high and low have different lengths"""
        high = np.array([10.0, 20.0, 30.0])
        low = np.array([5.0, 15.0])
        
        with pytest.raises(ValueError, match="High and low arrays must have the same length"):
            ta_indicators.kaufmanstop(high, low)
    
    def test_kaufmanstop_stream(self):
        """Test KaufmanstopStream class"""
        stream = ta_indicators.KaufmanstopStream(period=22, mult=2.0, direction="long", ma_type="sma")
        
        
        for i in range(21):
            result = stream.update(100.0 + i, 95.0 + i)
            assert result is None
        
        
        result = stream.update(122.0, 117.0)
        assert result is not None
        assert isinstance(result, float)
    
    def test_kaufmanstop_batch(self, test_data):
        """Test batch processing"""
        result = ta_indicators.kaufmanstop_batch(
            test_data['high'],
            test_data['low'],
            period_range=(20, 24, 2),  
            mult_range=(1.5, 2.5, 0.5),  
            direction="long",
            ma_type="sma"
        )
        
        
        assert 'values' in result
        assert 'periods' in result
        assert 'mults' in result
        assert 'directions' in result
        assert 'ma_types' in result
        assert 'rows' in result
        assert 'cols' in result
        
        
        assert result['rows'] == 9
        assert result['cols'] == len(test_data['high'])
        assert result['values'].shape == (9, len(test_data['high']))
        
        
        assert len(result['periods']) == 9
        assert len(result['mults']) == 9
        assert len(result['directions']) == 9
        assert len(result['ma_types']) == 9
    
    def test_kaufmanstop_kernel_options(self, test_data):
        """Test different kernel options"""
        kernels = [None, "scalar", "auto"]
        results = []
        
        for kernel in kernels:
            result = ta_indicators.kaufmanstop(
                test_data['high'], 
                test_data['low'],
                kernel=kernel
            )
            results.append(result)
        
        
        for i in range(1, len(results)):
            np.testing.assert_allclose(results[0], results[i], rtol=1e-10)
    
    def test_kaufmanstop_short_direction(self, test_data):
        """Test short direction"""
        result_long = ta_indicators.kaufmanstop(
            test_data['high'][:100], 
            test_data['low'][:100],
            direction="long"
        )
        
        result_short = ta_indicators.kaufmanstop(
            test_data['high'][:100], 
            test_data['low'][:100],
            direction="short"
        )
        
        
        assert not np.allclose(result_long, result_short, equal_nan=True)
        
        
        warmup = 21  
        for i in range(warmup, min(warmup + 10, 100)):
            if not np.isnan(result_long[i]) and not np.isnan(result_short[i]):
                
                assert result_long[i] != result_short[i], f"Long and short should differ at index {i}"
    
    def test_kaufmanstop_partial_params(self, test_data):
        """Test KAUFMANSTOP with partial parameters - mirrors check_kaufmanstop_partial_params"""
        
        result = ta_indicators.kaufmanstop(
            test_data['high'],
            test_data['low']
        )
        assert len(result) == len(test_data['high'])
    
    def test_kaufmanstop_default_candles(self, test_data):
        """Test KAUFMANSTOP with default parameters - mirrors check_kaufmanstop_default_candles"""
        
        expected = EXPECTED_OUTPUTS['kaufmanstop']
        result = ta_indicators.kaufmanstop(
            test_data['high'],
            test_data['low'],
            period=expected['default_params']['period'],
            mult=expected['default_params']['mult'],
            direction=expected['default_params']['direction'],
            ma_type=expected['default_params']['ma_type']
        )
        assert len(result) == len(test_data['high'])
    
    def test_kaufmanstop_very_small_dataset(self):
        """Test KAUFMANSTOP fails with insufficient data - mirrors check_kaufmanstop_very_small_dataset"""
        high = np.array([42.0])
        low = np.array([41.0])
        
        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.kaufmanstop(high, low, period=22)
    
    def test_kaufmanstop_empty_input(self):
        """Test KAUFMANSTOP fails with empty input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Empty data"):
            ta_indicators.kaufmanstop(empty, empty)
    
    def test_kaufmanstop_all_nan_input(self):
        """Test KAUFMANSTOP with all NaN values"""
        all_nan_high = np.full(100, np.nan)
        all_nan_low = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.kaufmanstop(all_nan_high, all_nan_low)
    
    def test_kaufmanstop_nan_handling(self, test_data):
        """Test KAUFMANSTOP handles NaN values correctly - mirrors check_kaufmanstop_nan_handling"""
        expected = EXPECTED_OUTPUTS['kaufmanstop']
        
        result = ta_indicators.kaufmanstop(
            test_data['high'],
            test_data['low'],
            period=expected['default_params']['period']
        )
        assert len(result) == len(test_data['high'])
        
        
        if len(result) > 240:
            assert not np.any(np.isnan(result[240:])), "Found unexpected NaN after index 240"
        
        
        warmup = expected['warmup_period']  
        assert np.all(np.isnan(result[:warmup])), f"Expected NaN in warmup period (first {warmup} values)"
    
    def test_kaufmanstop_streaming(self, test_data):
        """Test KAUFMANSTOP streaming matches batch calculation - mirrors check_kaufmanstop_streaming"""
        expected = EXPECTED_OUTPUTS['kaufmanstop']
        period = expected['default_params']['period']
        mult = expected['default_params']['mult']
        direction = expected['default_params']['direction']
        ma_type = expected['default_params']['ma_type']
        
        
        batch_result = ta_indicators.kaufmanstop(
            test_data['high'],
            test_data['low'],
            period=period,
            mult=mult,
            direction=direction,
            ma_type=ma_type
        )
        
        
        stream = ta_indicators.KaufmanstopStream(
            period=period,
            mult=mult,
            direction=direction,
            ma_type=ma_type
        )
        stream_values = []
        
        for i in range(len(test_data['high'])):
            result = stream.update(test_data['high'][i], test_data['low'][i])
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        
        assert len(batch_result) == len(stream_values)
        
        warmup = expected['warmup_period']
        for i in range(warmup, len(batch_result)):
            if np.isnan(batch_result[i]) and np.isnan(stream_values[i]):
                continue
            assert_close(batch_result[i], stream_values[i], rtol=1e-9, atol=1e-9,
                        msg=f"KAUFMANSTOP streaming mismatch at index {i}")
    
    def test_kaufmanstop_batch_single_params(self, test_data):
        """Test batch with single parameter set"""
        expected = EXPECTED_OUTPUTS['kaufmanstop']
        
        result = ta_indicators.kaufmanstop_batch(
            test_data['high'],
            test_data['low'],
            period_range=(22, 22, 0),  
            mult_range=(2.0, 2.0, 0.0),  
            direction="long",
            ma_type="sma"
        )
        
        assert 'values' in result
        assert result['values'].shape == (1, len(test_data['high']))
        
        
        single_row = result['values'][0]
        assert_close(
            single_row[-5:],
            expected['batch_default_row'],
            rtol=1e-6,
            msg="KAUFMANSTOP batch single params mismatch"
        )
    
    def test_kaufmanstop_batch_multiple_periods(self, test_data):
        """Test batch with multiple period values"""
        high = test_data['high'][:100]  
        low = test_data['low'][:100]
        
        result = ta_indicators.kaufmanstop_batch(
            high, low,
            period_range=(20, 24, 2),  
            mult_range=(2.0, 2.0, 0.0),
            direction="long",
            ma_type="sma"
        )
        
        
        assert result['values'].shape == (3, 100)
        assert result['rows'] == 3
        assert result['cols'] == 100
        
        
        periods = [20, 22, 24]
        for i, period in enumerate(periods):
            row_data = result['values'][i]
            single_result = ta_indicators.kaufmanstop(
                high, low,
                period=period,
                mult=2.0,
                direction="long",
                ma_type="sma"
            )
            np.testing.assert_allclose(row_data, single_result, rtol=1e-10)
    
    def test_kaufmanstop_batch_metadata(self, test_data):
        """Test batch result metadata"""
        high = test_data['high'][:50]
        low = test_data['low'][:50]
        
        result = ta_indicators.kaufmanstop_batch(
            high, low,
            period_range=(20, 22, 2),  
            mult_range=(1.5, 2.0, 0.5),  
            direction="long",
            ma_type="sma"
        )
        
        
        assert result['rows'] == 4
        assert result['cols'] == 50
        assert len(result['periods']) == 4
        assert len(result['mults']) == 4
        
        
        assert list(result['periods']) == [20, 20, 22, 22]
        assert list(result['mults']) == [1.5, 2.0, 1.5, 2.0]
    
    def test_kaufmanstop_different_ma_types(self, test_data):
        """Test various MA types"""
        
        ma_types = ["sma", "ema", "wma", "smma"]
        results = []
        
        for ma_type in ma_types:
            try:
                result = ta_indicators.kaufmanstop(
                    test_data['high'][:100],
                    test_data['low'][:100],
                    ma_type=ma_type
                )
                results.append((ma_type, result))
            except ValueError:
                
                pass
        
        
        assert len(results) >= 1
        
        
        
        unique_results = []
        for ma_type, result in results:
            is_unique = True
            for _, existing in unique_results:
                if np.allclose(result, existing, equal_nan=True):
                    is_unique = False
                    break
            if is_unique:
                unique_results.append((ma_type, result))
        
        
        assert len(unique_results) >= 1, "At least one MA type should produce results"


if __name__ == '__main__':
    pytest.main([__file__, '-v'])
